src/drivers/fluid_waveout.c

```diff
@@ -0,0 +1,359 @@
+        {
+            WAVEOUTCAPS caps;
+            MMRESULT    res;
+
+            res = waveOutGetDevCaps(n, &caps, sizeof(caps));
+            if (res == MMSYSERR_NOERROR)
+            {
+#ifdef _UNICODE
+                if(wcsicmp(lpwDevName, caps.szPname) == 0)
+#else
+                if(FLUID_STRCASECMP(dev_name, caps.szPname) == 0)
+#endif
+                {
+                    FLUID_LOG(FLUID_DBG, "Selected audio device GUID: %s", dev_name);
+                    device = n;
+                    break;
+                }
+            }
+        }
+    }
+
+    if(dev_name)
+    {
+        FLUID_FREE(dev_name);
+    }
+
+    /* Create thread which processes re-adding SYSEX buffers */
+    dev->hThread = CreateThread(
+                       NULL,
+                       0,
+                       (LPTHREAD_START_ROUTINE)
+                       fluid_waveout_synth_thread,
+                       dev,
+                       0,
+                       &dev->dwThread);
+
+    if(dev->hThread == NULL)
+    {
+        FLUID_LOG(FLUID_ERR, "Failed to create waveOut thread");
+        HeapFree(GetProcessHeap(), 0, dev);
+        return NULL;
+    }
+
+    errCode = waveOutOpen(&dev->hWaveOut,
+                          device,
+                          &wfx,
+                          (DWORD_PTR)dev->dwThread,
+                          0,
+                          CALLBACK_THREAD);
+
+    if (errCode != MMSYSERR_NOERROR)
+    {
+        FLUID_LOG(FLUID_ERR, "Failed to open waveOut device");
+
+        /* Close the thread */
+        PostThreadMessage(dev->dwThread, WM_CLOSE, 0, 0);
+        WaitForSingleObject(dev->hThread, INFINITE);
+        CloseHandle(dev->hThread);
+
+        HeapFree(GetProcessHeap(), 0, dev);
+        return NULL;
+    }
+
+    /* Get pointer to sound buffer memory */
+    ptrBuffer = (LPSTR)(dev+1);
+
+    /* Setup the sample buffers */
+    for (i = 0; i < NB_SOUND_BUFFERS; i++)
+    {
+        /* Clear the sample buffer */
+        memset(ptrBuffer, 0, lenBuffer);
+
+        /* Clear descriptor buffer */
+        memset(dev->waveHeader+i, 0, sizeof(WAVEHDR));
+
+        /* Compile descriptor buffer */
+        dev->waveHeader[i].lpData         = ptrBuffer;
+        dev->waveHeader[i].dwBufferLength = lenBuffer;
+
+        waveOutPrepareHeader(dev->hWaveOut, &dev->waveHeader[i], sizeof(WAVEHDR));
+
+        ptrBuffer += lenBuffer;
+    }
+
+    /* Play the sample buffers */
+    for (i = 0; i < NB_SOUND_BUFFERS; i++)
+    {
+        waveOutWrite(dev->hWaveOut, &dev->waveHeader[i], sizeof(WAVEHDR));
+    }
+
+    return (fluid_audio_driver_t *) dev;
+}
+
+
+void delete_fluid_waveout_audio_driver(fluid_audio_driver_t *d)
+{
+    int i;
+
+    fluid_waveout_audio_driver_t *dev = (fluid_waveout_audio_driver_t *) d;
+    fluid_return_if_fail(dev != NULL);
+
+    /* release all the allocated resources */
+    if (dev->hWaveOut != NULL)
+    {
+        waveOutReset(dev->hWaveOut);
+        waveOutClose(dev->hWaveOut);
```

Comment:
Why are you closing waveOut before unpreparing the headers? Shouldn't it be

```c
waveOutUnprepareHeader()
waveOutClose()
```

Reply:
Because I did a copy/paste mistake ^^;
Fixed.